### PR TITLE
Added relations

### PR DESCRIPTION
**Pull request by martinyde**
https://jira.itkdev.dk/browse/FORMS-740

- Added relations table.
- Added use of relations table
- Added command for importing into relations table.

### CODE REVIEW

File: os2forms_failed_jobs.info.yml [hunk 1]
```diff
@@ -6,4 +6,4 @@ core: 8.x
 core_version_requirement: ^8 || ^9
 dependencies:
   - 'webform:webform'
-  - 'drupal:advancedqueue'
+  - 'advancedqueue:advancedqueue'
```
Comment (rimi-itk) [inline os2forms_failed_jobs.info.yml:9]: 🎉

File: os2forms_failed_jobs.install [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+ * Code related to installing and updating this module.
+ */
+
+/**
+ * Implements hook_schema().
+ *
+ * @phpstan-return array<string, mixed>
+ */
+function os2forms_failed_jobs_schema(): array {
+  $schema['os2forms_queue_submission_relation'] = [
```
Comment (rimi-itk) [inline os2forms_failed_jobs.install]: Isn't it customary to prefix the table name with the module name, i.e. call the table `os2forms_failed_jobs_submission_jobs`, say?

File: os2forms_failed_jobs.install [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+}
+
+/**
+ * Add os2forms_queue_submission_relation table.
+ */
+function os2forms_failed_jobs_update_9001(): void {
+  $database = \Drupal::database();
+  $schema = $database->schema();
+
+  $table_name = 'os2forms_queue_submission_relation';
```
Comment (rimi-itk) [inline os2forms_failed_jobs.install]: See comment on `os2forms_failed_jobs_schema`.

File: os2forms_failed_jobs.services.yml [hunk 4]
```diff
@@ -1,4 +1,13 @@
...
-    arguments:
-      - '@database'
+    arguments: [
+      '@entity_type.manager',
+      '@database',
+      '@logger.factory'
+    ]
+
+  Drupal\os2forms_failed_jobs\EventSubscriber\AdvancedQueueEventsSubscriber:
+    class: Drupal\os2forms_failed_jobs\EventSubscriber\AdvancedQueueProcessSubscriber
```
Comment (rimi-itk) [inline os2forms_failed_jobs.services.yml]: The `class` property is superfluous when using the full class name as the service name.

File: src/Form/RetryJob.php [hunk 5]
```diff
@@ -52,22 +52,22 @@ class RetryJob extends ConfirmFormBase {
...
-    $this->failedJobsHelper = $failedJobsHelper;
+    $this->helper = $helper;
   }
 
   /**
    * {@inheritdoc}
    */
   public static function create(ContainerInterface $container): RetryJob {
-    return new static(
+    return new RetryJob(
```
Comment (rimi-itk) [inline src/Form/RetryJob.php]: Why do you change `static` to `RetryJob` here? It works the same, but `new static` is the way it's done in most (all?) cases.

File: src/Helper/Helper.php [hunk 6]
```diff
@@ -68,18 +93,198 @@ public function getSubmissionIdFromJob(int $jobId): ?int {
...
+   *
+   * @return string
+   *   The webform id.
+   */
+  public function getWebformIdFromQueue(string $jobId): string {
+    $query = $this->connection->select('os2forms_queue_submission_relation', 'o');
+    $query->condition('job_id', $jobId, '=');
+    $query->fields('o', ['webform_id']);
+
+    return $query->execute()->fetchObject()->webform_id;
```
Comment (rimi-itk) [inline src/Helper/Helper.php]: You should handle the case when the requested job id does not exist.

File: src/Helper/Helper.php [hunk 7]
```diff
@@ -68,18 +93,198 @@ public function getSubmissionIdFromJob(int $jobId): ?int {
...
+   *
+   * @param \Drupal\advancedqueue\Job $job
+   *   The job about to be processed.
+   */
+  public function handleJob(Job $job): void {
+    $data = $this->getDataFromPayload($job->getPayload());
+    if ($data) {
+      $data['job_id'] = (int) $job->getId();
+      $this->addRelation($data);
+    }
```
Comment (rimi-itk) [inline src/Helper/Helper.php]: I think the code will be a little cleaner if you rename `getDataFromPayload` to `getDataFromJob`, pass it the `Job` and have it return the full data (including `job_id`). 

Consider removing the `addRelation` function and move it's body into this function's body.  `addRelation` is only used in one place and is not too complex so putting the logic into `handleJob` will not make the code more complex.

File: src/Helper/Helper.php [hunk 8]
```diff
@@ -68,18 +93,198 @@ public function getSubmissionIdFromJob(int $jobId): ?int {
...
+
+    try {
+      return [
+        'submission_id' => $submissionId,
+        'webform_id' => $this->getWebformSubmission($submissionId)
+        ?->getWebform()
+        ?->id(),
+      ];
+    }
+    catch (InvalidPluginDefinitionException | PluginNotFoundException $e) {
```
Comment (rimi-itk) [inline src/Helper/Helper.php]: Are these exceptions ever thrown in the `try` block?

**Review by rimi-itk — CHANGES_REQUESTED** [review]

File: os2forms_failed_jobs.install (hunk 2 above)
Comment (martinyde) [reply os2forms_failed_jobs.install]: Changed

File: os2forms_failed_jobs.install (hunk 3 above)
Comment (martinyde) [reply os2forms_failed_jobs.install]: Changed

File: os2forms_failed_jobs.services.yml (hunk 4 above)
Comment (martinyde) [reply os2forms_failed_jobs.services.yml]: Removed property

File: src/Form/RetryJob.php (hunk 5 above)
Comment (martinyde) [reply src/Form/RetryJob.php]: Changed

File: src/Helper/Helper.php (hunk 7 above)
Comment (martinyde) [reply src/Helper/Helper.php]: Moved

File: src/Helper/Helper.php (hunk 8 above)
Comment (martinyde) [reply src/Helper/Helper.php]: Changed

File: src/Helper/Helper.php (hunk 6 above)
Comment (martinyde) [reply src/Helper/Helper.php]: Fixed using null safe operator

**Review by rimi-itk — APPROVED** [review]
We could ignore PHP 8.0 (which is almost “End of life”) and go for just 8.1.

**Pull request merged**